### PR TITLE
shell: remove Console dependencies

### LINKED ISSUE #10191: The new shell uses CONSOLE kconfig options, even though it does not use CONSOLE.
The new shell uses CONSOLE kconfig options, even though it does not use CONSOLE.

Related to #10190 & #10127

### PR DESCRIPTION
**Pull request by jakub-uC**
Removed Console dependencies from shell uart backend.
Generated define: CONFIG_UART_SHELL_ON_DEV_NAME for each board.

Fixes #10191

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

### CODE REVIEW

**Comment by jarz-nordic** [comment]
@aescolar : Please let me know what I need to update in regards to POSIX to pass CI tests?
I see these tests are failing.

**Review by aescolar — DISMISSED** [review]
It is not just the posix boards, but any board which is not yet using DTS.
You need to add to the shell Kconfig file something like this one:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/console/Kconfig#L45

Say in subsys/shell/Kconfig.backends : 26 add:

```
if !HAS_DTS
config UART_SHELL_ON_DEV_NAME
	string "Device Name of UART Device for SHELL_BACKEND_SERIAL"
	default "UART_0"
	help
	  This option specifies the name of UART device to be used for the
	  SHELL UART backend.
endif
```

**Comment by jarz-nordic** [comment]
@aescolar : Thank you very much for hint. I will take care to fix it.

**Review by pizi-nordic — APPROVED** [review]

**Comment by jarz-nordic** [comment]
@aescolar : May you please take another look at failed CI?
I am having segmentation fault from cmake.

**Comment by aescolar** [comment]
@jarz-nordic @ulfalizer It would seem the crash is in the kconfig python script (?)

**Comment by jarz-nordic** [comment]
@ulfalizer : May you please take a look at failing CI?

**Comment by ulfalizer** [comment]
@jarz-nordic 
No idea to be honest.

Looks like a segfault when running `scripts/kconfig/kconfig.py`, like @aescolar said, but Python normally doesn't segfault (stack overflows and the like are caught and raise exceptions instead).

Maybe it's running out of some resource, or there's something wonky with the Python installation or test environment. Guessing the bug isn't in the script at least.

**Comment by ulfalizer** [comment]
Small Kconfig tip btw:

Putting `if X` around a single symbol is the same as putting `depends on X` on the symbol itself, so could have `depends on !HAS_DTS`.

Not sure it's any clearer tbh, but `if` is just a shorthand for adding a `depends on` to everything within the `if` at least. There are no "parse-time" `if`s in Kconfig (would get tricky, because you can have forward references).

**Review by carlescufi — APPROVED** [review]

**Comment by jarz-nordic** [comment]
@nashif :
Anas, as expected PR is failing due to timeout (too much tests to be executed). Can we merge it as is?
It is hard for me to track each and every new board added to Zephyr to update this PR.

**Comment by aescolar** [comment]
@jarz-nordic Is that mysterious crash accounted for? (or at least not reproducible after reasonable attempts?)
[Kind of scary to merge something that is timeouting today when in the previous run it went down in flames]

**Comment by jarz-nordic** [comment]
@aescolar : I've discussed it with @ulfalizer offline. He has no explanation why it crashed but he suspected:
1. messed-up installation
2. server went out of resources
and he has recommended to re-run the CI.

**Review by pfalcon — APPROVED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
Looks good to me, but we need to probably improve the docs related to the Zephyr chosen properties from a user perspective.

**Comment by carlescufi** [comment]
>Looks good to me, but we need to probably improve the docs related to the Zephyr chosen properties from a user perspective.
@galak 
Do you know where those are documented?

**Comment by pabigot** [comment]
I'm not in love with putting this in the DTS as a selected hardware resouce: it smells like more of an application-determined preference.

Can the dependency be managed with a Kconfig option that defaults to `"CONSOLE"` which selects the existing `zephyr,console` value but can be overridden with another device name under the assumption that the device tree already makes sure that device is also available?

**Comment by pfalcon** [comment]
@pabigot : I'm sure @jarz-nordic will reply better, but what you see is a result of a series of negotiations and compromises to get forward with this, even if that's one incremental step, to be adjusted later. And we really need to go forward with this.

If you're interested, some (not all) discussion is captured in https://github.com/zephyrproject-rtos/zephyr/issues/10512

**Comment by pabigot** [comment]
@pfalcon OK.   It was my understanding we want to reduce the amount of Zephyr-specific content in DTS, and on the surface the added property seems redundant.  If it touches more than just the binding name that's referenced in only one source file it's not that simple.  Just wanted to verify it'd been considered.

**Comment by carlescufi** [comment]
@galak @jarz-nordic maybe [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/devices/dts/device_tree.rst) we could document Zephyr's use of the `chosen` keyword?

**Comment by pfalcon** [comment]
@pabigot: My understanding that it's done like that, following how it's done for "console" already. And the reason to split it away from "console" is because shell and console kinda compete for the underlying resource (e.g. UART). And that "console" is quite adhoc notion in Zephyr, some people (e.g. me) would like to make it to be much more POSIX-inspired, but it's not yet there. Once it's there, it's a hope that we can rejoin shell with console, i.e. make it run on top of a console, and "everything else" too - all that subject to verification of requirements at the point when console is actually "ready". So, the current move is just to partition them nicely, to avoid confusing interdependencies.

**Review by galak — CHANGES_REQUESTED** [review]
I think we should have something in:

https://docs.zephyrproject.org/latest/application/application.html

and also probably a page similar to (or part of):

https://docs.zephyrproject.org/latest/reference/kconfig/index.html

That documents all the 'chosen' properties we support in Zephyr.

**Review by galak — APPROVED** [review]
LGTM, maybe let @dbkinder take a look at the docs.

File: doc/devices/dts/device_tree.rst [hunk 1]
```diff
@@ -248,6 +248,68 @@ The following is a more precise list of required files:
     :ref:`application_dt` for more information on overlay files and the Zephyr
     build system.
 
+.. _dt-alias-chosen:
+
+``aliases`` and ``chosen`` nodes
+================================
+
+It is often useful to alias a particular node to a common name that can be
```
Comment (dbkinder) [inline doc/devices/dts/device_tree.rst]: For clarity, how about changing the first sentence to:
```
Using an alias with a common name for a particular node makes
it easier for you to write board-independent source code.
```

File: doc/devices/dts/device_tree.rst [hunk 2]
```diff
@@ -248,6 +248,68 @@ The following is a more precise list of required files:
...
+
+   aliases {
+      led0 = &led0;
+      sw0 = &button0;
+      sw1 = &button1;
+      uart-0 = &uart0;
+      uart-1 = &uart1;
+   };
+
+Certain software subsystems typically require a specific hardware resource to
```
Comment (dbkinder) [inline doc/devices/dts/device_tree.rst]:
```suggestion
Certain software subsystems require a specific hardware resource to
```

File: doc/devices/dts/device_tree.rst [hunk 3]
```diff
@@ -248,6 +248,68 @@ The following is a more precise list of required files:
...
+bind to in order to function properly. Some of those subsystems are typically
```
Comment (dbkinder) [inline doc/devices/dts/device_tree.rst]:
```suggestion
bind to in order to function properly. Some of those subsystems are
```

File: doc/devices/dts/device_tree.rst [hunk 4]
```diff
@@ -248,6 +248,68 @@ The following is a more precise list of required files:
...
+used with many different boards, which makes using the Device Tree ``chosen``
+nodes very convenient. By doing so the software subsystem can rely on having
```
Comment (dbkinder) [inline doc/devices/dts/device_tree.rst]:
```suggestion
nodes very convenient. By doing so, the software subsystem can rely on having
```

File: doc/devices/dts/device_tree.rst [hunk 5]
```diff
@@ -248,6 +248,68 @@ The following is a more precise list of required files:
...
+The full set of Zephyr-specific ``chosen`` nodes follows:
+
+.. list-table::
+   :header-rows: 1
+
+   * - ``chosen`` node name
+     - Generated symbol
+
+   * - ``zephyr,flash``
+     - ``CONFIG_FLASH``
```
Comment (dbkinder) [inline doc/devices/dts/device_tree.rst]: any reason you didn't use ``:option:`CONFIG_FLASH` `` here (and on the next two items?)

**Review by dbkinder — DISMISSED** [review]

File: doc/devices/dts/device_tree.rst (hunk 5 above)
Comment (jarz-nordic) [reply doc/devices/dts/device_tree.rst]: @dbkinder 
These options not exist in Kconfig so we want to keep it as non-options.

**Comment by jarz-nordic** [comment]
@dbkinder : All your findings were addressed, please review again.

**Comment by carlescufi** [comment]
Merging this since comments are addressed and this affects many boards, so letting it linger makes it complicated to maintain it. @dbkinder if you find an issue post-merge please ping us and we'll fix it.

**Comment by dbkinder** [comment]
@carlescufi doc changes LGTM (comments were addressed)

**Pull request merged**